src/libraries/System.Private.CoreLib/src/System/IO/File.cs

```diff
@@ -1007,5 +1007,40 @@ public static Task AppendAllLinesAsync(string path, IEnumerable<string> contents
                 ? Task.FromCanceled(cancellationToken)
                 : InternalWriteAllLinesAsync(AsyncStreamWriter(path, encoding, append: true), contents, cancellationToken);
         }
+
+        /// <summary>
+        /// Creates a file symbolic link identified by <paramref name="path"/> that points to <paramref name="pathToTarget"/>.
+        /// </summary>
+        /// <param name="path">The path where the symbolic link should be created.</param>
+        /// <param name="pathToTarget">The path of the target to which the symbolic link points.</param>
+        /// <returns>A <see cref="FileInfo"/> instance that wraps the newly created file symbolic link.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="path"/> or <paramref name="pathToTarget"/> is <see langword="null"/>.</exception>
+        /// <exception cref="ArgumentException"><paramref name="path"/> or <paramref name="pathToTarget"/> is empty.
+        /// -or-
+        /// <paramref name="path"/> is not an absolute path.
+        /// -or-
+        /// <paramref name="path"/> or <paramref name="pathToTarget"/> contains invalid path characters.</exception>
```

Comment:
Grouping `path` and `pathToTarget` here makes it sound like they will both be validated against the same set of invalid characters.  `pathToTarget` allows almost anything, including most of `Path.InvalidPathChars`.

Reply:
@carlreinke, a symlink points to another file-system entry using a regular path, so it seems sensible to validate the target path _formally_ like any other path (not in terms of existence; although, given that symlinks can span file-systems, conceivably they have different sets of valid characters - how does `FileSystem.VerifyValidPath` handle this currently?). Why do you think `pathToTarget` allows (should allow) almost anything?

Reply:
@mklement0 `pathToTarget` allows almost anything because `mklink` (and thus the underlying OS API) allows almost anything in the target on Windows.
```
C:\temp>mklink foo "<>:/\|?*"
symbolic link created for foo <<===>> <>:/\|?*

C:\temp>dir | findstr foo
2021-06-16  23:08    <SYMLINK>      foo [<>:/\|?*]
```
(On Linux the only invalid path character is `'\0'`, so it's obvious that it will allow almost anything there.)

See [this comment](https://github.com/dotnet/runtime/issues/24271#issuecomment-830341663) for a use case for storing arbitrary data in a symlink.

Reply:
@carlreinke: 

Wow - I certainly did not expect `mklink foo "<>:/\|?*"` to work, and you have to wonder why that is permitted. The implementation via reparse points means that arbitrary data can be stored, so one can see how it's _technically_ possible, but to what end? Probably _not_ in anticipation of "off-label" uses such as as a lock on Unix, per the comment you link to.

I guess that disallowing `'\0'` is then the only check we can perform without making underlying capabilities inaccessible (I haven't tried the APIs directly, but at least `mklink` / `ln` don't support `'\0'`.)

The only caveat is that we must make sure that other .NET APIs then don't run into trouble when they encounter  a `FileSystemInfo` instance with such a misshapen path obtained via `.ResolveLinkTarget()` - note that
`new FileInfo("<>:/\|?*")` does _not_ work, for instance.

And, of course, this permissiveness should be clearly documented.

Reply:
De-facto only OS file system driver can do full path validation. In the library it makes sense only to do simple checks for null/`\0`.